### PR TITLE
Switch to thrift server with multiple selector threads.

### PR DESCRIPTION
**Pull request by keith-turner**
Before this change Accumulo's used the THsHaServer thrift server. This
server has a single select threads that read and writes frames to
sockets.  The select thread puts the frames on a thread pool for
processing.  On a machine with a lot of cores, this single thread can
become a bottleneck.  This commit switches to use the thrift
TThreadedSelectorServer server which support multiple select threads.

To demonstrate the differernce this change can make, the
RandomCachedLookupsPT performance test in accumulo-testing was run with
and without this change.  The test was run on a machine with 16 cores.
The test loads a table into cache and then does random lookups with
varying numbers of threads.  It measures the average lookup time for
each lookup. Below is a table of running this test.

| Threads | Single Selector | Multi Selector | Speedup |
| --------|-----------------|----------------|---------|
|     1 | 0.14 ms         | 0.14 ms        | 1.00 |
|      4 | 0.18 ms         | 0.19 ms        | 0.95 |
|      8 | 0.24 ms         | 0.24 ms        | 1.00 |
|     16 | 0.44 ms         | 0.44 ms        | 1.00 |
|     32 | 0.85 ms         | 0.81 ms        | 1.05 |
|     64 | 1.74 ms         | 1.40 ms        | 1.24 |
|    128 | 3.51 ms         | 2.44 ms        | 1.44 |
|    256 | 7.21 ms         | 4.52 ms        | 1.60 |

This table shows that with 256 client threads doing lookups that each
lookup took 7.21ms on average with a single selector thread.  With this
change and four selector threads, each loopup took 4.52ms on average
with 256 client threads.  During both test there were 256 threads to
process client request on the tserver.  The selector threads hand frames
off to these 256 threads.

Watching top+jstack while running these test show the single selector
thread will eventually reach 100% on a core.  With 4 selector threads
each used 36% CPU when 256 client threads were doing lookups.

### CODE REVIEW

File: server/base/src/main/java/org/apache/accumulo/server/rpc/ThriftServerType.java [hunk 1]
```diff
@@ -28,7 +28,11 @@
  * benchmarks against "unsecure" Accumulo use the same type of Thrift server.
  */
 public enum ThriftServerType {
-  CUSTOM_HS_HA("custom_hs_ha"), THREADPOOL("threadpool"), SSL("ssl"), SASL("sasl");
+  CUSTOM_HS_HA("custom_hs_ha"),
+  THREADPOOL("threadpool"),
+  SSL("ssl"),
+  SASL("sasl"),
+  THREADED_SELECTOR("threaded_selector");
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/rpc/ThriftServerType.java]: This is probably going to rejoin on formatting.

File: server/base/src/main/java/org/apache/accumulo/server/rpc/CustomThreadedSelectorServer.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+      TNonblockingTransport transport = getTransport(frameBuffer);
+
+      if (transport instanceof TNonblockingSocket) {
+        // This block of code makes the client address available to the server side code that
+        // executes a RPC. It is made available for informational purposes.
+        TNonblockingSocket tsock = (TNonblockingSocket) transport;
+        Socket sock = tsock.getSocketChannel().socket();
+        TServerUtils.clientAddress
+            .set(sock.getInetAddress().getHostAddress() + ":" + sock.getPort());
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/rpc/CustomThreadedSelectorServer.java]: This is stored statically? Is it stored in a ThreadLocal?
Also, I hate how we have to copy huge chunks of libthrift code just to log a client address.

File: server/base/src/main/java/org/apache/accumulo/server/rpc/CustomThreadedSelectorServer.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to access required field in Thrift code.", e);
+    }
+  }
+
+  private TNonblockingTransport getTransport(FrameBuffer frameBuffer) {
+    try {
+      return (TNonblockingTransport) fbTansportField.get(frameBuffer);
+    } catch (IllegalArgumentException | IllegalAccessException e) {
+      throw new RuntimeException(e);
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/rpc/CustomThreadedSelectorServer.java]: I know this is copied from Thrift, but still seems unnecessary to wrap an RTE with a new RTE.

File: server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java [hunk 4]
```diff
@@ -186,7 +187,43 @@ public static ServerAddress startServer(ServerContext service, String hostname,
...
+      TProcessor processor, TProtocolFactory protocolFactory, final String serverName,
+      final int numThreads, final int numSTThreads, long timeBetweenThreadChecks,
+      long maxMessageSize) throws TTransportException {
+
+    final TNonblockingServerSocket transport = new TNonblockingServerSocket(
+        new InetSocketAddress(address.getHost(), address.getPort()));
+
+    TThreadedSelectorServer.Args options = new TThreadedSelectorServer.Args(transport);
+
+    options.selectorThreads = Math.max(2, Runtime.getRuntime().availableProcessors() / 4);
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java]: Could make this tunable in future.

File: server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java [hunk 5]
```diff
@@ -186,7 +187,43 @@ public static ServerAddress startServer(ServerContext service, String hostname,
   }
 
   /**
-   * Create a NonBlockingServer with a custom thread pool that can dynamically resize itself.
+   * Create a non blocking server with multiple select threads and a custom thread pool that can
+   * dynamically resize itself.
+   */
+  public static ServerAddress createThreadedSelectorServer(HostAndPort address,
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java]: Does this server type support SSL or SASL? If so, we could consolidate server implementations a bit.

**Comment by ctubbsii** [comment]
Are there any scenarios where this server could conceivably be significantly worse?

File: server/base/src/main/java/org/apache/accumulo/server/rpc/CustomThreadedSelectorServer.java (hunk 3 above)
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/rpc/CustomThreadedSelectorServer.java]: It's not copied from thrift, it just a little method I created to do the reflection stuff.  Eclipse did that automatically I wasn't paying close attention.

File: server/base/src/main/java/org/apache/accumulo/server/rpc/ThriftServerType.java (hunk 1 above)
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/rpc/ThriftServerType.java]: I didn't change it, its just what the formatter decided to do.

File: server/base/src/main/java/org/apache/accumulo/server/rpc/CustomThreadedSelectorServer.java (hunk 2 above)
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/rpc/CustomThreadedSelectorServer.java]: Yeah, it's a static thread local.  It is not technically a copy, its an extension with a single method overridden. This one is much shorter than the CustomNonBlockingServer it superseded. It's possible that CustomNonBlockingServer could be made a lot shorter.

File: server/base/src/main/java/org/apache/accumulo/server/rpc/CustomThreadedSelectorServer.java (hunk 3 above)
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/rpc/CustomThreadedSelectorServer.java]: I removed the catch of IlegalArg

File: server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java (hunk 5 above)
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/rpc/TServerUtils.java]: I have no idea.

**Comment by keith-turner** [comment]
> Are there any scenarios where this server could conceivably be significantly worse?

None that I know of, however its certainly a possibility. That is why I left the old one the code, should be able to configure it if necessary.

**Pull request merged**